needy/cd.py

```diff
@@ -2,15 +2,26 @@
 
 _current_directory = None
 
+
 def current_directory():
     """ returns the current directory as given to cd (which means unresolved symlinks) """
     global _current_directory
+
+    ret = ''
+
     if _current_directory:
-        return _current_directory
-    try:
-        return os.environ['PWD']
-    except:
-        return os.getcwd()
+        ret = _current_directory
+    else:
+        try:
+            ret = os.environ['PWD']
+        except:
+            ret = os.getcwd()
+
+    if ret is None or os.path.realpath(ret) is not os.path.realpath(os.getcwd()):
+        ret = os.getcwd()
+
+    return ret
```

Comment:
`return ret if os.path.realpath(ret) is os.path.realpath(os.getcwd()) else os.getcwd()`